RecoJets/JetProducers/plugins/VirtualJetProducer.cc

```diff
@@ -846,6 +856,114 @@ void VirtualJetProducer::writeCompoundJets(  edm::Event & iEvent, edm::EventSetu
+    p4_Jets.push_back( math::XYZTLorentzVector(localJet.px(), localJet.py(), localJet.pz(), localJet.e() ));
+    double localJetArea = 0.0;
+    if ( doAreaFastjet_ && localJet.has_area() ) {
+      localJetArea = localJet.area();
+    }
+    area_Jets.push_back( localJetArea );
+
+    // create the constituent list
+    std::vector<fastjet::PseudoJet> constituents,ghosts;
+    if ( it->has_pieces() )
+      constituents = it->pieces();
+    else if ( it->has_constituents() )
+      fastjet::SelectorIsPureGhost().sift(it->constituents(), ghosts, constituents); //filter out ghosts
+
+    //loop over constituents of jet (can be subjets or normal constituents)
+    for (fastjet::PseudoJet const& constit : constituents) {
+      if ( verbosity_ >= 1 ) {
+        std::cout << "jet #" << jetIndex <<
+          ": Pt = " << constit.pt() <<
+          ", eta = " << constit.eta() <<
+          ", phi = " << constit.phi() <<
+          ", mass = " << constit.m() <<
+          ", uid: " << constit.user_index() <<
+          ", pos: " << constituentsSub.size() <<
+          ")" << std::endl;
+      }
+      indices[jetIndex].push_back( constituentsSub.size() );
+      constituentsSub.push_back(constit);
+    }
+  }
+
+  //Loop over constituents and store in the event
+  static const reco::PFCandidate dummySinceTranslateIsNotStatic;
+  for (fastjet::PseudoJet const& constit : constituentsSub) {
+    auto orig = inputs_[constit.user_index()];
+    auto id = dummySinceTranslateIsNotStatic.translatePdgIdToType(orig->pdgId());
+    reco::PFCandidate pCand( reco::PFCandidate(orig->charge(), orig->p4(), id) );
+    math::XYZTLorentzVector pVec;
+    pVec.SetPxPyPzE(constit.px(),constit.py(),constit.pz(),constit.e());
+    pCand.setP4(pVec);
+    pCand.setSourceCandidatePtr( orig->sourceCandidatePtr(0) );
+    constituentCollection->push_back(pCand);
+  }
+  // put constituents into event record
+  constituentHandleAfterPut = iEvent.put(std::move(constituentCollection), jetCollInstanceName_ );
+
+  // Now create the jets with ptr's to the constituents
+  std::vector<math::XYZTLorentzVector>::const_iterator ip4 = p4_Jets.begin(),
+    ip4Begin = p4_Jets.begin(),
+    ip4End = p4_Jets.end();
+
+  for ( ; ip4 != ip4End; ++ip4 ) {
+    int p4_index = ip4 - ip4Begin;
+    std::vector<int> & ind = indices[p4_index];
+    std::vector<reco::CandidatePtr> i_jetConstituents;
+    // Add the constituents to the jet
+    for( std::vector<int>::const_iterator iconst = ind.begin(); iconst != ind.end(); ++iconst ) {
+      reco::CandidatePtr candPtr( constituentHandleAfterPut, *iconst, false );
+      i_jetConstituents.push_back( candPtr );
+    }
+    if(i_jetConstituents.size()>0) { //only keep jets which have constituents after subtraction
+      reco::Particle::Point point(0,0,0);
+      reco::PFJet jet;
+      reco::writeSpecific(jet,*ip4,point,i_jetConstituents,iSetup);
+      jet.setJetArea( area_Jets[ip4 - ip4Begin] );
+      jetCollection->push_back( jet );
```

Comment:
I changed it to emplace_back and checked that it doesn't change the result. I however lack experience to understand the difference properly between push_back and emplace_back. After googling a bit it seems emplace_back can be used by calling the constructor directly but that is not possible in this case since we have to set the area.